contracts/adapters/midas/MidasIssuanceVaultAdapter.sol

```diff
@@ -0,0 +1,148 @@
+
+    /// @dev Set of allowed input tokens for depositInstant
+    EnumerableSet.AddressSet internal _allowedTokens;
+
+    /// @notice Constructor
+    /// @param _creditManager Credit manager address
+    /// @param _midasIssuanceVault Midas Issuance Vault address
+    /// @param _referrerId Referrer ID to use for all deposits
+    constructor(address _creditManager, address _midasIssuanceVault, bytes32 _referrerId)
+        AbstractAdapter(_creditManager, _midasIssuanceVault)
+    {
+        mToken = IMidasIssuanceVault(_midasIssuanceVault).mToken();
+
+        // We check that mToken is a valid collateral
+        _getMaskOrRevert(mToken);
+
+        referrerId = _referrerId;
+    }
+
+    /// @notice Deposits specified amount of input token for mToken
+    /// @param tokenIn Input token address
+    /// @param amountToken Amount of input token to deposit
+    /// @param minReceiveAmount Minimum amount of mToken to receive
+    function depositInstant(address tokenIn, uint256 amountToken, uint256 minReceiveAmount, bytes32)
+        external
+        override
+        creditFacadeOnly
+        returns (bool)
+    {
+        if (!isTokenAllowed(tokenIn)) revert TokenNotAllowedException();
+
+        _depositInstant(tokenIn, amountToken, minReceiveAmount);
+        return false;
+    }
+
+    /// @notice Deposits entire balance of input token, except the specified amount
+    /// @param tokenIn Input token address
+    /// @param leftoverAmount Amount of input token to keep in the account
+    /// @param rateMinRAY Minimum exchange rate from input token to mToken (in RAY format)
+    function depositInstantDiff(address tokenIn, uint256 leftoverAmount, uint256 rateMinRAY)
+        external
+        override
+        creditFacadeOnly
+        returns (bool)
+    {
+        if (!isTokenAllowed(tokenIn)) revert TokenNotAllowedException();
+
+        address creditAccount = _creditAccount();
+
+        uint256 balance = IERC20(tokenIn).balanceOf(creditAccount);
+        if (balance > leftoverAmount) {
+            unchecked {
+                uint256 amount = balance - leftoverAmount;
+                uint256 minReceiveAmount = (amount * rateMinRAY) / RAY;
+                _depositInstant(tokenIn, amount, minReceiveAmount);
+            }
+        }
+        return false;
+    }
+
+    /// @dev Internal implementation of `depositInstant`
+    function _depositInstant(address tokenIn, uint256 amountToken, uint256 minReceiveAmount) internal {
+        _executeSwapSafeApprove(
+            tokenIn,
+            abi.encodeCall(
+                IMidasIssuanceVault.depositInstant,
+                (tokenIn, _convertToE18(amountToken, tokenIn), minReceiveAmount, referrerId)
+            )
+        );
+    }
+
+    /// @dev Converts the token amount to 18 decimals, which is accepted by Midas
+    function _convertToE18(uint256 amount, address token) internal view returns (uint256) {
+        uint256 tokenUnit = 10 ** IERC20Metadata(token).decimals();
```

Comment:
Can probably early return if `decimals` equals 18 to avoid unnecessary exponentiation